ext/drwrap/drwrap.h

```diff
@@ -88,34 +88,41 @@ drwrap_exit(void);
 DR_EXPORT
 /**
  * When #drwrap_global_flags_t #DRWRAP_INVERT_CONTROL is set, the user
- * must call this function from a drmgr analysis event handler
+ * must call this function from a drmgr insertion event handler
  * (typically registered with
  * drmgr_register_bb_instrumentation_event() or if using drbbdup with
- * the analyze_case_ex field in #drbbdup_options_t).  It is up to the user
- * to control the ordering, since the priority #DRMGR_PRIORITY_INSERT_DRWRAP
- * will not apply.
+ * the instrument_instr_ex field in #drbbdup_options_t).  This
+ * function will insert instrumentation for function wrapping pre and
+ * post callbacks.  It is up to the user to control the ordering,
+ * since the priority #DRMGR_PRIORITY_INSERT_DRWRAP will not apply.
+ * The separate "where" handles cases such as with drbbdup's final app
+ * instruction (which cannot be duplicated into each case) or with
+ * emulation where the instruction "inst" to monitor is distinct from
+ * the location "where" to insert instrumentation.
  */
 dr_emit_flags_t
-drwrap_invoke_analysis(void *drcontext, void *tag, instrlist_t *bb, bool for_trace,
-                       bool translating, OUT void **user_data);
+drwrap_invoke_insert(void *drcontext, void *tag, instrlist_t *bb, instr_t *inst,
+                     instr_t *where, bool for_trace, bool translating, void *user_data);
 
 DR_EXPORT
 /**
- * When #drwrap_global_flags_t #DRWRAP_INVERT_CONTROL is set, the user
- * must call this function from a drmgr insertion event handler
- * (typically registered with
- * drmgr_register_bb_instrumentation_event() or if using drbbdup with
- * the instrument_instr_ex field in #drbbdup_options_t).  It is up to the user
- * to control the ordering, since the priority #DRMGR_PRIORITY_INSERT_DRWRAP
- * will not apply.
+ * When #drwrap_global_flags_t #DRWRAP_INVERT_CONTROL is set and the
+ * user is using multiple types of instrumentation such as with
+ * drbbdup, the user must call this function for each instrumentation
+ * case where function wrapping should not be enabled.  This will
+ * insert required cleanup for instrumentation cases changing in the
+ * middle of a wrapped function.  It will not invoke any wrapped
+ * function callbacks.  It is up to the user to control the ordering,
+ * since the priority #DRMGR_PRIORITY_INSERT_DRWRAP will not apply.
  * The separate "where" handles cases such as with drbbdup's final app
  * instruction (which cannot be duplicated into each case) or with
  * emulation where the instruction "inst" to monitor is distinct from
  * the location "where" to insert instrumentation.
  */
 dr_emit_flags_t
-drwrap_invoke_insert(void *drcontext, void *tag, instrlist_t *bb, instr_t *inst,
-                     instr_t *where, bool for_trace, bool translating, void *user_data);
+drwrap_invoke_insert_cleanup_only(void *drcontext, void *tag, instrlist_t *bb,
```

Comment:
So the cleanup code is added always, even in the drbbdup instrumentation case where drwrap didn't wrap. I see that this works fine when the drbbdup case gets changed in the middle of a wrapped function; it allows us to cleanup the changes we made to wrap the function. But, is it okay to cleanup if we didn't wrap at all?

Reply:
Yes, the drwrap post-call handling checks whether it's in a target function or not.  It has to, to handle some other control flow reaching the same basic block that a retaddr targets.